README
9/9/2012
* use node.js connect()
* change logic for startM2()

9/9/2012
We're using a virtual box and run the web server (node.js) in it. The M2's are started in schroots (secure chroots)

9/6/2012
We considered jails in FreeBSD briefly, but decided to use Ubuntu schroots.

8/31/2012
* get it running on Mike's newest machine
* maybe get MacOS 10 Server, can we sandbox it on there? 
* get sandbox running on Lars' laptop

* Security issues: We are now using a startup script setting ulimits. Can these be overwritten from inside the schroot?
* upload a package
* upload tutorials 

* use it in the cloud with more than 1 machine

* Write a paper, maybe JSAG?





4/20/2012
* problem: 2 tabs in the same browser share the same cookie and the same eventStream
* what happens when we quit the browser? 

Timer: 
- start timer when user leaves
- destroy timer when user with same cookie comes back
- Can we determine if M2 is currently running a calculation for a client?



Mike: get server ready for May visit

4/17/2012
Do not start a new user every time they leave the page:  DONE
- don't delete clientID right away from clients[] (DONE), use a timer instead  
- Done: we are currently writing to many (dead) clients with ondata, we need to "unlink" previous on data calls first, before assigning them to new eventStreams

Still to do: 
display image not in help box  DONE
cronjob to cleanup /tmp (Maybe Lars?)
/tmp should be only writable directory in sandbox
make images work in sandbox and still have security (talk to Dan about allowing allowing restricted M2 to write)

maybe work on Print and Don't reload when briefly navigating away from site
Can we run a non-restricted M2 inside the sandbox and still be secure?


4/3/2012
source .profile because open is in export PATH=/Users/franzi/open-bin:$PATH
goal for today: serve the right image, not a static one



3/22/2012
cleanup loadFiles for jpg
get PID - find user by PID


3/20/12
app-armor?

allowing the display of created jpg or png images (e.g. with Graphs.m2).
  one possible design:
    a. M2 places all such files into a specific directory, based on the pid (in chroot dir)


  problem: the 'open' script which would communicate with the server, doesn't know who created
    then picture. [it could do: curl localhost:8000/picture/$pid/a.jpg]
      a. use getppid
      b. does curl
     m2server: handles /image/, findClientIDFromPID, gets actual file name (in sandbox), 
       sends on the event stream for that client, an object of type image:
     m2node: on client side, ondata routine needs to check the 'image' type, and if it is set, it loads the file
       in a overlayed window (or whatever).


Gwyn: 
ability to get worksheet uploaded
print
Dan: 2 dir's for packages: experimental
users can start their own server
want to be able to run programs e.g., gfan...
be able to display jpg files and generate them as in posets package
list of allowed commands  (like gfan etc)
don't delete history on reload in input window
ask David Cook about M2/gap
test low memory conditions, gracefully telling the user 
lots of users -> issues? 
Save input and output windows
sage - notebook doesn't run M2
link for reporting issues, asking questions

3/14/2012
bug: website is unresponsive for minutes (?) when the connection was closed in between
"keyboard" with buttons for (){}_#... to make iPad use easier


3/13/2012
multi-line <code>


Greg: 
"useful starting code" in Terminal
Get rid of M2 title, put in welcome screen instead
Bug: same browser, different tabs, several M2's in the same output window
help 'command' should open up in tutorial-style css, maybe on the left side? 
Tutorial button on the bottom of the page? 
syntax highlighting and autocompletion (maybe a button not necessarily tab)
Previous and Next should be inside the tutorial screen
re-arrange title, buttons, menu 

3/12/2012
  In general: security
    includes: not serving pages in /etc/...!
              denial of service attack
              running security screening tools (nlmp, or...?)
              https protocol, more secure cookies/session  MOSTLY DONE

  streamlining the M2 schroot:
    limit resources  almost DONE
    ...
  
  server mechanics:
    cron job to restart server, if it dies
    logging of data, who is coming (ip address, OS?)
    kill extraneous M2 processes
    learn about load balancing and reverse proxy server
    remove stale M2's  (time out: after a week, ...)
  server itself:
    at the moment: serving "/" resends index.hmtl and restarts M2.
    would be nice if it didn't do that: it could remember process if cookie is OK.
  
  user authentication
    allowing files on the server? Better persistence, better M2 performance (less limits)

3/7/2012
For version 1: See Mike's list


3/6/12
* deal with viewHelp, how can we display help to users? 
* Navigating through lessons, scroll bar should jump to top of each lesson

2/29/2012
* email Greg
* talk to Lars next week

2/23/2012
- Talking about server pushing data to client rather than client pulling for data

2/21/2012
Get "Next" button working again! - Done
Node.js, chat server example in "Definite Guide to Javascript", Flanagan, p 519, we need Lars to incorporate this into the existing server

Homework due 2/21 - 80% of your final grade!!!
Franzi: convert BeginningMacaulay2 to Tutorial (trunk/Macaulay2/packages/Begin...)
Mike: investigate asynchronous retrieval of M2 output, JS lint, read up on "global" variables in JS

2/17/12
At this point, it's basically functional, we need to clean it up a little bit: 
	- add another tutorial
	- fix help text
	- credits (funding?)
	- generally clean code, get rid of obsolete (?) createMenu() and createMenu2()
	- display title of current tutorial
	- display 
	
	
In the future:	
* we need 
	- load lessons from url
* write content for help text
* need Greg make it prettier
* get rid of constant pulling of data
* iPad (buttons instead of keyboard input)
* save/get our work (via email)
* Stylized way of making a tutorial page, conversion between M2 documentation style and tutorial form 
* resizing
* communicate the text-area size to M2 for correct string formatting or set reasonable default at beginning, printWidth = 80, but figure out what the width should be in advance
* undo feature	( command Z?)
* syntax highlighting
* autocompletion on Tab
* save contents (input and output)
* upload files that will turn into clickable tutorials, notebook style
* Run all commands in the tutorial up to here


2/7/12
* MB Extruder, http://pupunzi.open-lab.com/mb-jquery-components/jquery-mb-extruder/
* we need 
	- show tutorial lessons
	- load lessons
	- load lessons from url
* Strip down extruder files (js and css in root dir), change way menu is getting its content (we have a function for retrieving content, i.e., all H4's, in m2.js, createMenu()
	
	
1/30/12
* final prompts in output terminal
* AJAX-like call to get results from M2 when they're ready instead of pulling every x ms, get a smarter web server
* a little CSS to make it just a little prettier (colors, font, buttons, shadows, offset inside boxes, don't type into output window)
* indexing for tutorials, from here, load tutorials
* second tutorial so we can play with double indexed list
* help button


11/9/11
Adjusting layout, using http://tour.golang.org/ as template


12/10/11
Giving up on Sencha for now (chapter2, sencha*.*)
Going back to index-ipad.html, renaming it to index.html, starting fresher




Sencha Touch: 

* textarea size
* node.js for "pulling" data from the server without draining the battery
* Ajax calls (this should be really easy - really?)
* more buttons
* larger textarea
* lessons
* we want swiping
* table of contents/lesson overview



Use index.html in main directory or index-ipad.html, start the server from inside sockets: 
ruby RubyM2Server.rb

The php scripts that you're using from inside the Javascript files is sockets/M2Client.php and getResults.php. Be careful with different directories ...

-------

* clicking on "send to M2" button should move the cursor to next line
* padding inside lessons

* fix TOC on help page (!)
* fix TOC expand/collapse
* fix the layout :) 
* change M2 output to not be a text area but something with scrollbars
* sendToM2 functional on lesson pages, send everything up to that point
* table of contents: collapse/expand, fix sizes
* help button
* tutorial button
* back button change name
* somehow select font size

* colorization of keywords, comments, and strings


Done :) 
* detect orientation, side by side, over each other
* change titles: "on the web", "interactive" -> remove or change
* take out waiting time
* 2 finger zoom

**Planned Feature List** (some things might be out of date...)
resizing
incorporate input from tutorial
communicate the text-area size to M2 for correct string formatting or set reasonable default at beginning
undo feature	( command Z?)
syntax highlighting
autocompletion on Tab
save contents (input and output)
upload files that will turn into clickable tutorials, notebook style
Run all commands in the tutorial up to here
retrieve newest results right after sending a request and then less frequently

iPad: 
* scroll bars
* math keyboard, or buttons of most needed keys
* move text areas around
* newRing button: k, p, gens
* buttons for variables that are available

Involve Amelia, Frank Moore, Bart Snapp, other undergrad research people to have undergrads use this site, maybe develop lessons, develop a project that can be used with this site? Brandy (Elena?) thought computational algebra, any insight?



**Tutorial**
change structure: Preamble and div's should be in index.html, just load commands (and explanations)
Choice of tutorials to chose from
"run all" button, run selection/line


Use index.html in main directory or index-ipad.html, start the server from inside sockets: 
ruby RubyM2Server.rb

The php scripts that you're using from inside the Javascript files is sockets/M2Client.php and getResults.php. Be careful with different directories ... 


-----

in the cloud
education: for undergrads
why not sage: old version, too complex for simple in-classroom action, 

notebook structure not right format for M2
not suitable for mobile version
"join" sage: our notebooks can be made available through sage as well, for a different kind of users

colorization


releasenotes.txt

2.0.0 6/23/2017
    * Markdown format for wirting tutorials
    * Internally: use TypeScript

1.1.0 4/13/2015
    * Multi-line code tags
    * Docker containers instead of schroots
    * Vagrantfile for users who want to host interactive shell
    * Tab completion on Macaulay2 commands
    * New layout
    * Public-common and public-specific
    * Users can upload tutorials
    * Singular support

1.0.0 3/31/2013
Stable version running since March 2013
    * shell-like right hand side
    * upload and download of files
    * supports graphs.m2 package
    * several basic tutorials




m2-node.js
/*global $, SyntaxHighlighter, alert, console, document, trym2, updateOrientation, window */


var trym2 = {
    lessonNr: 1,
    maxLesson: 1
};

trym2.scrollDown = function (area) {
    var mySize = $(area).val().length;
    $(area).scrollTop(mySize);
    return false;
    // Return false to cancel the default link action
};

/* get selected text, or current line, in the textarea #M2In */
trym2.getSelected = function (inputField) {
    var str = $(inputField).val(),
        start = $(inputField)[0].selectionStart,
        end = $(inputField)[0].selectionEnd,
        endPos;
    if (start === end) {
        // grab the current line
        start = 1 + str.lastIndexOf("\n", end - 1);
        endPos = str.indexOf("\n", start);
        if (endPos !== -1) {
            end = endPos;
        } else {
            end = str.length;
        }
    }
    return str.slice(start, end) + "\n";
};

trym2.showTerminal = function () {
    $("#lesson").hide();
    $("#inputarea").show();
    $("#send").show();
    $("#pageIndex").hide();
    $("#previous").hide();
    $("#next").hide();
    $("#showLesson").show();
    $("#terminal").hide();
    return false;
};

trym2.loadLesson = function (ell) {
    $("#inputarea").hide();
    var lessonContent = $('[lessonid="' + ell + '"]').html();
    $("#send").hide();
    $("#previous").show();
    $("#next").show();
    $("#pageIndex").text(trym2.lessonNr + "/" + trym2.maxLesson).show();
    $("#lesson").html(lessonContent).show();
    $("#showLesson").hide();
    $("#terminal").show();
};

trym2.switchLesson = function (incr) {
    //console.log("Current lessonNr " + trym2.lessonNr);
    //console.log("maxlesson " + trym2.maxLesson);
    trym2.lessonNr = trym2.lessonNr + incr;
    if (trym2.lessonNr >= 1 && trym2.lessonNr <= trym2.maxLesson) {
        //console.log("Switch lesson");
        trym2.loadLesson(trym2.lessonNr);
        $("#lesson").scrollTop(0);
    } else {
        trym2.lessonNr = trym2.lessonNr - incr;
        //alert("lesson with " + trym2.lessonNr + "." + incr + " not available");
    }
};

trym2.callback = function( url, msg ) {
    return function() {
        var xhr = new XMLHttpRequest();           // Create a new XHR
        //console.log( "URL: " + url);
        xhr.open("POST", url);                // to POST to url.
        xhr.setRequestHeader("Content-Type",      // Specify plain UTF-8 text 
                            "text/plain;charset=UTF-8");
        xhr.onreadystatechange = function() {
            if ( xhr.readyState === 4) {
                //console.log( "All ResponseHeaders: " + xhr.getAllResponseHeaders());
                var resHead = xhr.getResponseHeader('notEventSourceError');
                //console.log( "ResponseHeader: " + resHead);
                if (resHead){
                    console.log("We must have lost the EventSource Stremt, redoing it...");
                    trym2.startEventSource();
                    //send msg again
                    trym2.callback("/chat", msg)();
                }
            }
        };
        xhr.send(msg);                            // Send the message        
        return true;
    }  
};

trym2.sendCallback = function (inputField) {
    return function () {
        var str = trym2.getSelected(inputField);
        trym2.callback('/chat', str)();
        return false;
    };
};

trym2.sendOnEnterCallback = function (inputfield) {
    return function (e) {
        if (e.which === 13 && e.shiftKey) {
            e.preventDefault();
            // do not make a line break or remove selected text when sending
            trym2.callback('/chat',trym2.getSelected(inputfield))();
        }
    };
};

trym2.helpScreen = function () {
    //console.log("Display Help.");
    $("#help-dialog").dialog('open');
    $("#help-dialog").scrollTop(0);
    
    
};

// input: filename with tutorial content
// return a list of lessons title, each wrapped in a div and link, 
// <div><a>" + title + "</a></div>  
// attach a lesson ID
trym2.getLessonTitles = function (tutorialFile, callback) {
    var titles = "";
    $("#menuTutorial").load(tutorialFile, function () {
        var i = 1;
        $("#menuTutorial h4").each(function () {
            var title = $(this).text();
            titles = titles + "<div><a class='submenuItem' lessonid='lesson" + i + "'>" + title + "</a></div>";
            i = i + 1;
            //console.log("Title in m2.js: " + title);
        });
        //console.log("All titles: " + titles);
        callback(titles);
    });
};

// Register for notification of new messages using EventSource
trym2.startEventSource = function () {
    var chat = new EventSource("/startSourceEvent");
    chat.addEventListener('image', function(event) {
        var imageUrl = event.origin + event.data;
        console.log("Image coming from: " + event.origin);
        //console.log("We got an image! " + imageUrl);
        if (imageUrl) {
            console.log("We got an image! " + imageUrl);
            $('<div title="Graph"><img src="' + imageUrl + '"/><div>').dialog({
                height: 340,
                width: 460,
                modal: true,
                autoOpen: true
            });
            //$("#help-dialog").dialog('open');
            //$("#help-dialog").scrollTop(0);
            //$('#help').click(trym2.helpScreen);
            //$("#help-dialog").prepend('<h1>Fancy Image!</h1><img src="' + imageUrl + '"/>');
            //trym2.helpScreen();
        }
    }, false);
    chat.onmessage = function(event) {            // When a new message arrives
         var msg = event.data;                     // Get text from event object
         console.log(event);
         
         if (msg !== "") {
                 //console.log("We got a chat message: " + msg);
                 $("#M2Out").val($("#M2Out").val() + msg);
                 trym2.scrollDown("#M2Out");
         }

     }
};

$(document).ready(function () {
    // send server our client.eventStream
    trym2.startEventSource();
    var extruderMenu = $("#extruderTop").buildMbExtruder({
        position: "top",
        width: 350,
        extruderOpacity: 1,
        onExtOpen: function () {},
        onExtContentLoad: function () {},
        onExtClose: function () {}
    });
    
    $('#M2In').val("Evaluate a line by typing Shift+Enter or by clicking on Evaluate.\nHere are some sample commands:\n---------------\nR = ZZ/101[a,b,c]\nS = ZZ/32003[vars(1..10)]\nQQ[x_1..x_6]\n");
    
    //$('#M2In').addClass('inactive');
    //$('#M2In').focus(function() {
    //    $(this).removeClass('inactive'); 
    //    if ($(this).val() == "Write an M2 command, e.g., 3+3, and hit Shift+Enter or click on Evaluate.") {
    //        $(this).val('');
    //    }
    //});
    //trym2.startEventSource();
    $('.submenuItem').live("click", function () {
        var i = 1,
            lessonId = $(this).attr('lessonid');
        //console.log("You clicked a submenuItem: " + $(this).html());
        trym2.lessonNr = parseInt(lessonId.match(/\d/g), 10);
        $("#tutorial").html($("#menuTutorial").html());
        $("#tutorial h4").each(function () {
            $(this).parent().attr('lessonid', i); // add an ID to every lesson div
            i = i + 1;
        });
        trym2.maxLesson = i-1;
        trym2.loadLesson(trym2.lessonNr);
    });

    $('#help-dialog').dialog({
        height: 340,
        width: 460,
        modal: true,
        autoOpen: false
    });
    $('#help').click(trym2.helpScreen);

    SyntaxHighlighter.all();
    
    $("#send").click(trym2.sendCallback('#M2In'));
    $('#M2In').keypress(trym2.sendOnEnterCallback('#M2In'));
    $("#reset").click(trym2.callback('/restart'));
    $("#interrupt").click(trym2.callback('/interrupt'));
    $("#terminal").click(trym2.showTerminal);
    $("#inputTerminalLink").live( "click", trym2.showTerminal);
    
    $("#showLesson").click(function() {
        trym2.loadLesson(trym2.lessonNr);
        //console.log("lesson!");
    });
  

    $("code").live("click", function () {
        $(this).effect("highlight", {color: 'red'}, 300);
        var code = $(this).text();
        code = code + "\n";
        $("#M2In").val($("#M2In").val() + code);
        trym2.scrollDown("#M2In");
        trym2.callback('/chat', code)();
    });

    $("#inputarea").hide();
    $("#send").hide();

    $("#pageIndex").hide();

$("#tutorial").html("<div class='lesson' lessonid='1'>  <div><br>Get started by <a href='#' id='selectTutorialLink'><b>selecting a tutorial</b></a> or by using the <a href='#' id='inputTerminalLink'><b>Input Terminal</b></a>. Have fun!<br> <p><code> needsPackage \"Graphs\"<br>\nA = graph({{x_1,x_3},{x_2,x_4},{x_1,x_4}})<br>\ndisplayGraph A</code><p><code>4+3*9</code><br><code>apply(10, i -> i *17)</code><br><code>R = ZZ/3[a,b,c]</code><br><code>groebnerBasis ideal(a^2, a*b-b*b, -c^2)</code><br><a href='http://www.math.uiuc.edu/Macaulay2/'>Macaulay2</a> is a software system devoted to supporting research in algebraic geometry and commutative algebra, whose creation has been funded by the National Science Foundation since 1992. </p> <p> To get started, select a tutorial. Click on any highlighted code, Macaulay2 will execute it. The result is displayed on the right. Alternatively, you can use the <b>Input Terminal</b> on the left to write your own commands. Execute a line by positioning your cursor on it and click on the Evaluate button (or type Shift-Enter). You can switch back to the tutorial at any time. </p> <p>The tutorials demonstrate different aspects of Macaulay2. They are meant to be starting points for your own experimentation. Edit the commands in the <b>Input Terminal</b> and run them again. Whenever you're ready to move on, click the Next button.</p>");
// <code>3+18</code><br>    <code>version</code><br>    <code> exit </code> <br>   <code> R = ZZ/101[vars(0..13)] </code> <br><code> gbTrace=1 </code> <br> <code> time res coker vars R </code> <br>  <code> m1 = genericMatrix(R,a,3,3) </code> <br> <code> m2 = genericMatrix(R,j,3,3) </code> <br> <code> J = ideal(m1*m2-m2*m1) </code> <br> <code> C = res J </code> <br> <code> C.dd_3 </code> <br> 
    
    $("#selectTutorialLink").live("click", function() {
        //console.log("open tutorial menu");
        extruderMenu.openMbExtruder();
        return false;
    });
    
    trym2.loadLesson(trym2.lessonNr);
    trym2.maxLesson = $('#tutorial').children().length;
    //console.log("maxLesson: " + trym2.maxLesson);

    $("#next").click(function () {
        trym2.switchLesson(1);
    });
    $("#previous").click(function () {
        trym2.switchLesson(-1);
    });
    
    $("#next").hide();
    $("#previous").hide();
    $("#showLesson").hide();

    // swipe changed to swipeXXX to remove functionality for testing
    $(function () {
        $("#leftwindow").bind("swipeXXX", function (event, info) {
            if (info.direction === "left") {
                trym2.switchLesson(1);
            } else if (info.direction === "right") {
                trym2.switchLesson(-1);
            } else {
                alert("swiped: huh?");
            }
        });
    });



    //updateOrientation();
});

function updateOrientation() {
    var orient = "";
    switch (window.orientation) {
    case 0:
    case 180:
        orient = "show_portrait";
        break;
    case -90:
    case 90:
        orient = "show_landscape";
        break;
    default:
        orient = "show_landscape";
    }
    $("body").attr("class", orient);
    $("#rightwindow").attr('class', orient);
    $("#leftwindow").attr('class', orient);
}

